docs/rustdoc/prehashing.md

```diff
@@ -0,0 +1,191 @@
+- The probability that a tuple of full size is selected, $q$, is inversely proportional to $d$ and approximates $\lambda_{rel} / d$:
+  - Since $d$ is related to $u$, $q$ is also approximately inversely proportional to $u$.
+  - A smaller $q$ improves security by lowering the chance of accepting cheating proofs but decreases the probability of finding valid proofs, requiring a larger $d$ to compensate.
+
+$$
+u \coloneqq \Bigg\lceil \frac{\lambda_{sec} + \log (\lambda_{rel} + \log 3) + 1 - \log \log e}{\log (n_p / n_f)}\Bigg\rceil,
+$$
+$$
+d \coloneqq \Bigg\lceil \frac{16 u (\lambda_{rel} + \log 3)}{\log e}\Bigg\rceil , \quad q \coloneqq \frac{2 (\lambda_{rel} + \log 3)}{d \log e}.
+$$
+Furthermore, to maintain a high probability of successfully constructing a valid proof, $n_p$ must satisfy:
+$$
+n_p \geq \frac{d^2 \log e}{9(\lambda_{rel} + \log 3)}.
+$$
+- This condition ensures that the prover's set is large enough to achieve a well-distributed hashing of elements into bins.
+- If $n_p$ is too small, bins may not be sufficiently populated, reducing the chances of constructing valid tuples.
+
+### Example walkthrough
+- **Prover’s set** $S_p = \{A, B, C, D\}$
+- **Proof length** $u = 3$
+- **Search width** $d = 2$ (i.e., $t \in \{1, 2\}$)
+- **Hash functions**:
+  - $H_0(s)$: Prehash for binning elements of $S_p$.
+  - $H_1(v, t, s_1, \ldots, s_k)$: Determines the next bin to explore based on the sequence so far.
+  - $H_2(v, t, s_1, \ldots, s_u)$: Final validity check for the sequence.
+> ---
+> - The prover precomputes $H_0(s)$ for all elements in $S_p$, assigning each element to a bin:
+>   - $H_0(A) = 2$
+>   - $H_0(B) = 1$
+>   - $H_0(C) = 2$
+>   - $H_0(D) = 3$
+> - Bins are grouped as follows:
+>   - Bin 1: $\{B\}$
+>   - Bin 2: $\{A, C\}$
+>   - Bin 3: $\{D\}$
+> ---
+> - The prover starts with $t = 1$ and attempts to construct a valid sequence of length $u = 3$ using bounded DFS.
+> - Depth $k = 1$:
+>   - Compute $H_1(1)$ to determine the first bin to explore.
+>     - Assume $H_1(1) = 2$, so the prover looks in bin 2.
+>   - Select $s_1 = A$ from bin 2.
+>     - Compute $H_1(1, A)$ to check where to explore next.
+>     - $H_1(1, A)$ takes the current sequence $(1, A)$ and outputs the bin index for the next step.
+>     - Assume $H_1(1, A) = 3$, meaning the prover should explore bin 3.
+> - Depth $k = 2$:
+>   - Look in bin 3, which contains $\{D\}$.
+>   - Select $s_2 = D$.
+>   - Compute $H_1(1, A, D)$ to check where to explore next.
+>     - Assume $H_1(1, A, D) = 1$, meaning the prover should explore bin 1.
+> - Depth $k = 3$ (Final Element):
+>   - Look in bin 1, which contains $\{B\}$.
+>   - Select $s_3 = B$.
+>   - Compute $H_1(1, A, D, B)$ to check if this sequence can proceed further.
+>     - Assume $H_1(1, A, D, B) = 3$, which indicates a mismatch with the required condition. The sequence is invalid.
```

Comment:
We shouldn't have this check. Now that we have u=3 elements, we directly check with H2=1

Reply:
```suggestion
>     - Assume $H_2(1, A, D, B) = 0$, which indicates a mismatch with the required condition. The sequence is invalid.
```